pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+
+    if TYPE_CHECKING:
+        MessageQueue = asyncio.Queue[Tuple[SubscriberMessage,  # pylint: disable=unsubscriptable-object
+                                           float]]
+    else:
+        MessageQueue = asyncio.Queue
+    ApplicationHandler = Callable[[SubscriberMessage], Awaitable[None]]
+
+    class AckDeadlineCache:
+        def __init__(self, subscriber_client: SubscriberClient,
+                     subscription: str, cache_timout: int):
+            self.subscriber_client = subscriber_client
+            self.subscription = subscription
+            self.cache_timeout = cache_timout
+            self.ack_deadline: float = float('inf')
+            self.last_refresh: float = 0
+
+        async def get(self) -> float:
+            if self.ack_deadline == float('inf') or self.cache_outdated():
+                await self.refresh()
+            return self.ack_deadline
+
+        async def refresh(self) -> None:
+            try:
+                sub = await self.subscriber_client.get_subscription(
+                    self.subscription)
+                self.ack_deadline = float(sub['ackDeadlineSeconds'])
+            except Exception as e:
+                log.warning(
+                    'Failed to refresh ackDeadlineSeconds value', exc_info=e)
+            self.last_refresh = time.perf_counter()
+
+        def cache_outdated(self) -> bool:
+            if (time.perf_counter() - self.last_refresh) > self.cache_timeout:
+                return True
+            return False
+
+    async def acker(subscription: str,
+                    ack_queue: 'asyncio.Queue[str]',
+                    subscriber_client: 'SubscriberClient',
+                    ack_window: float,
+                    metrics_client: Optional[MetricsAgent]) -> None:
+        ack_ids: List[str] = []
+        while True:
+            time_budget = ack_window
+            if not ack_ids:
+                ack_ids.append(await ack_queue.get())
+                ack_queue.task_done()
+
+            while time_budget > 0:
+                start = time.perf_counter()
+                try:
+                    ack_id = await asyncio.wait_for(
+                        ack_queue.get(), timeout=time_budget)
+                    ack_ids.append(ack_id)
+                    ack_queue.task_done()
+                except asyncio.TimeoutError:
+                    break
+                time_budget -= (time.perf_counter() - start)
+            try:
+                await subscriber_client.acknowledge(subscription,
+                                                    ack_ids=ack_ids)
+            except Exception as e:
+                log.warning(
+                    'Ack request failed, better luck next batch', exc_info=e)
+                continue
+
+            if metrics_client:
+                metrics_client.histogram('pubsub.acker.batch', len(ack_ids))
+
+            ack_ids = []
+
+    async def _fire_callback(semaphore: asyncio.Semaphore,
```

Comment:
@allandialpad we were discussing naming here.

What I tried to capture here is the idea that we fire those tasks and pretty much forget about them. `task_wrapper` and `callback_wrapper` are a bit too generic for me. What others think?

Reply:
executor?

Reply:
Executor could be reasonable, so long as we can avoid confusing it with the asyncio executor concept: https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor . Should be fine if we're careful on naming.

Reply:
maybe `execute_callback`? Would be nice to keep it a verb since all surrounding "doer" functions are workers